react/components/ProgressBar/utils.ts

```diff
@@ -50,7 +56,14 @@ export function isDelivered(packages: any) {
   }
 
   let isDelivered = true
-  packages.map((pack: any) => {
+
+  var outputPackages = packages.filter(
+    (pack: any) => pack?.package?.type === OutputPackageType
+  )
+
+  if (outputPackages.length === 0) return false
+
+  outputPackages.map((pack: any) => {
     if (
       !pack.package ||
       !pack.package.courierStatus ||
```

Comment:
I think you could improve performance of this solution using the `some` function instead of `filter` and `map`.

I created a performance test to check it and using `some` looks like the best speed than the current solution. You can see the test using [this link](https://perf.link/#eyJpZCI6InB1ejB4cTZjMHZnIiwidGl0bGUiOiJpc0RlbGl2ZXJlZCIsImJlZm9yZSI6ImNvbnN0IHBhY2thZ2VzID0gW1xuICB7XG4gICAgcGFja2FnZToge1xuICAgICAgdHlwZTogXCJPdXRwdXRcIixcbiAgICAgIGNvdXJpZXJTdGF0dXM6IHtcbiAgICAgICAgZmluaXNoZWQ6IGZhbHNlLFxuICAgICAgfSxcbiAgICB9LFxuICB9LFxuICB7XG4gICAgcGFja2FnZToge1xuICAgICAgdHlwZTogXCJ0ZXN0XCIsXG4gICAgICBjb3VyaWVyU3RhdHVzOiB7XG4gICAgICAgIGZpbmlzaGVkOiB0cnVlLFxuICAgICAgfSxcbiAgICB9LFxuICB9LFxuICB7XG4gICAgcGFja2FnZToge1xuICAgICAgdHlwZTogXCJ0ZXN0XCIsXG4gICAgICBjb3VyaWVyU3RhdHVzOiB7XG4gICAgICAgIGZpbmlzaGVkOiB0cnVlLFxuICAgICAgfSxcbiAgICB9LFxuICB9LFxuICB7XG4gICAgcGFja2FnZToge1xuICAgICAgdHlwZTogXCJ0ZXN0XCIsXG4gICAgICBjb3VyaWVyU3RhdHVzOiB7XG4gICAgICAgIGZpbmlzaGVkOiBmYWxzZSxcbiAgICAgIH0sXG4gICAgfSxcbiAgfSxcbiAge1xuICAgIHBhY2thZ2U6IHtcbiAgICAgIHR5cGU6IFwidGVzdFwiLFxuICAgICAgY291cmllclN0YXR1czoge1xuICAgICAgICBmaW5pc2hlZDogdHJ1ZSxcbiAgICAgIH0sXG4gICAgfSxcbiAgfSxcbiAge1xuICAgIHBhY2thZ2U6IHtcbiAgICAgIHR5cGU6IFwidGVzdFwiLFxuICAgICAgY291cmllclN0YXR1czoge1xuICAgICAgICBmaW5pc2hlZDogdHJ1ZSxcbiAgICAgIH0sXG4gICAgfSxcbiAgfSxcbiAge1xuICAgIHBhY2thZ2U6IHtcbiAgICAgIHR5cGU6IFwidGVzdFwiLFxuICAgICAgY291cmllclN0YXR1czoge1xuICAgICAgICBmaW5pc2hlZDogZmFsc2UsXG4gICAgICB9LFxuICAgIH0sXG4gIH0sXG4gIHtcbiAgICBwYWNrYWdlOiB7XG4gICAgICB0eXBlOiBcInRlc3RcIixcbiAgICAgIGNvdXJpZXJTdGF0dXM6IHtcbiAgICAgICAgZmluaXNoZWQ6IHRydWUsXG4gICAgICB9LFxuICAgIH0sXG4gIH0sXG4gIHtcbiAgICBwYWNrYWdlOiB7XG4gICAgICB0eXBlOiBcInRlc3RcIixcbiAgICAgIGNvdXJpZXJTdGF0dXM6IHtcbiAgICAgICAgZmluaXNoZWQ6IHRydWUsXG4gICAgICB9LFxuICAgIH0sXG4gIH0sXG5dO1xuXG5jb25zdCBPdXRwdXRQYWNrYWdlVHlwZSA9ICdPdXRwdXQnIiwidGVzdHMiOlt7Im5hbWUiOiJQUiIsImNvZGUiOiIgIGxldCBpc0RlbGl2ZXJlZCA9IHRydWVcblxuICB2YXIgb3V0cHV0UGFja2FnZXMgPSBwYWNrYWdlcy5maWx0ZXIoXG4gICAgKHBhY2spID0%2BIHBhY2s%2FLnBhY2thZ2U%2FLnR5cGUgPT09IE91dHB1dFBhY2thZ2VUeXBlXG4gIClcblxuICBpZiAob3V0cHV0UGFja2FnZXMubGVuZ3RoID09PSAwKSByZXR1cm4gZmFsc2VcblxuICBvdXRwdXRQYWNrYWdlcy5tYXAoKHBhY2spID0%2BIHtcbiAgICBpZiAoXG4gICAgICAhcGFjay5wYWNrYWdlIHx8XG4gICAgICAhcGFjay5wYWNrYWdlLmNvdXJpZXJTdGF0dXMgfHxcbiAgICAgIHBhY2sucGFja2FnZS5jb3VyaWVyU3RhdHVzLmZpbmlzaGVkID09PSBmYWxzZVxuICAgICkge1xuICAgICAgaXNEZWxpdmVyZWQgPSBmYWxzZVxuICAgICAgcmV0dXJuXG4gICAgfVxuICB9KVxuXG4gIGNvbnNvbGUubG9nKGlzRGVsaXZlcmVkKSIsInJ1bnMiOls0ODAwMCw0OTAwMCwxMDAwLDEyMDAwLDUzMDAwLDM5MDAwLDEwMDAsOTIwMDAsMTE5MDAwLDE2NjAwMCwzMDAwMCwxNDAwMCwxMDAwLDkwMDAwLDc5MDAwLDE1NjAwMCwxMTYwMDAsNTAwMCw0OTAwMCwxNDAwMCw0MjAwMCw3NDAwMCwxNDAwMCwxNDAwMCwxNDAwMCw3NzAwMCwzODAwMCwyOTAwMCw5OTAwMCwxMDAwLDY1MDAwLDM5MDAwLDEwMDAsMTA0MDAwLDE0MjAwMCwxMjMwMDAsMTIzMDAwLDYyMDAwLDExMjAwMCwyNDAwMCw2MjAwMCw1MzAwMCwxMDAwLDEwMDAsMTA0MDAwLDE2NTAwMCwyNDAwMCw5MDAwLDQwMDAwLDEyODAwMCwxODAwMCw0OTAwMCw4ODAwMCw2NjAwMCwxMTUwMDAsNTAwMCwxMDUwMDAsMTAwMDAsOTAwMDAsMCw4NzAwMCw0MDAwMCwxMDAwLDUwMDAsMTAwMDAsMTUxMDAwLDMzMDAwLDg5MDAwLDMyMDAwLDE0ODAwMCw4OTAwMCw5NDAwMCwxMzgwMDAsNTAwMCw2MzAwMCw1NzAwMCw2ODAwMCwyMDUwMDAsMzQwMDAsNTYwMDAsODQwMDAsODAwMCwyOTAwMCw5MDAwLDE3NTAwMCwxMjIwMDAsNzgwMDAsMTA4MDAwLDEwNjAwMCwxMDEwMDAsMTAwMCwxMTcwMDAsMTUzMDAwLDE2MjAwMCwxMDUwMDAsNzUwMDAsMTAwMCwxMDcwMDAsNzcwMDAsMzAwMDBdLCJvcHMiOjY1MTcwfSx7Im5hbWUiOiJTdWdnZXN0aW9uIiwiY29kZSI6IiAgbGV0IGlzRGVsaXZlcmVkID0gdHJ1ZVxuXG4gIHBhY2thZ2VzLm1hcCgocGFjaykgPT4ge1xuICAgIGlmIChcbiAgICAgIHBhY2s%2FLnBhY2thZ2U%2FLnR5cGUgPT09IE91dHB1dFBhY2thZ2VUeXBlICYmIFxuICAgICAgKCFwYWNrLnBhY2thZ2UgfHxcbiAgICAgICFwYWNrLnBhY2thZ2UuY291cmllclN0YXR1cyB8fFxuICAgICAgcGFjay5wYWNrYWdlLmNvdXJpZXJTdGF0dXMuZmluaXNoZWQgPT09IGZhbHNlKVxuICAgICkge1xuICAgICAgaXNEZWxpdmVyZWQgPSBmYWxzZVxuICAgICAgcmV0dXJuXG4gICAgfVxuICB9KVxuXG4gIGNvbnNvbGUubG9nKGlzRGVsaXZlcmVkKSIsInJ1bnMiOlsxNjAwMCwzNDAwMCw2MDAwLDE2MDAwLDYxMDAwLDEwMDAsMTE4MDAwLDkxMDAwLDYwMDAwLDUxMDAwLDEwMjAwMCwxNTMwMDAsNDIwMDAsMzYwMDAsNzIwMDAsMTQzMDAwLDMwMDAwLDgwMDAsNTUwMDAsMjYwMDAsMzYwMDAsOTAwMCwzMDAwMCw3NDAwMCwxMzAwMCw3MTAwMCwxMTAwMCwzNzAwMCw3NzAwMCwxMjcwMDAsNzAwMDAsMTkwMDAsMTY1MDAwLDY4MDAwLDE2ODAwMCwxMTIwMDAsMjE2MDAwLDUxMDAwLDExNDAwMCwzMTAwMCwxMzYwMDAsMTkwMDAsMTA2MDAwLDEwMDAsMTA5MDAwLDMzMDAwLDU0MDAwLDk4MDAwLDIxMDAwLDcwMDAsMzMwMDAsNTAwMDAsNDMwMDAsMjIwMDAsMTA3MDAwLDYwMDAsMTU0MDAwLDg0MDAwLDE1MzAwMCw3ODAwMCw4NjAwMCwyMjAwMCwxNDQwMDAsMTQ1MDAwLDEzOTAwMCwxMjYwMDAsNjAwMDAsMTM4MDAwLDExMDAwLDE1MDAwMCw4MDAwMCw2MzAwMCwzMjAwMCw5NjAwMCw0NjAwMCw0MDAwLDEwMDAsNjYwMDAsNjAwMCw2OTAwMCwzMzAwMCwxMDAwLDg0MDAwLDgwMDAsNjgwMDAsNzcwMDAsMTEzMDAwLDIwMDAsMTAzMDAwLDU4MDAwLDEyNzAwMCwxMjUwMDAsOTUwMDAsMTA3MDAwLDE3ODAwMCw5MDAwMCwxMjMwMDAsODgwMDAsMTAwMCwxMDAwXSwib3BzIjo2OTAwMH0seyJuYW1lIjoiU29tZSIsImNvZGUiOiIgIGNvbnN0IGlzRGVsaXZlcmVkID0gcGFja2FnZXMuc29tZShcbiAgICAocGFjaykgPT5cbiAgICAgIHBhY2s%2FLnBhY2thZ2U%2FLnR5cGUgPT09IE91dHB1dFBhY2thZ2VUeXBlICYmXG4gICAgICAoIXBhY2sucGFja2FnZSB8fFxuICAgICAgICAhcGFjay5wYWNrYWdlLmNvdXJpZXJTdGF0dXMgfHxcbiAgICAgICAgcGFjay5wYWNrYWdlLmNvdXJpZXJTdGF0dXMuZmluaXNoZWQgPT09IGZhbHNlKVxuICApO1xuXG5cbiAgY29uc29sZS5sb2coaXNEZWxpdmVyZWQpIiwicnVucyI6Wzc5MDAwLDYwMDAsMzEwMDAsMjIwMDAsNjcwMDAsMTE5MDAwLDk0MDAwLDExNjAwMCwyOTAwMCw0NTAwMCw1ODAwMCw1MzAwMCwxNzAwMCwyMjAwMCwxMjAwMCwxODUwMDAsODMwMDAsOTAwMCwzMTAwMDAsNzMwMDAsMTYwMDAsNzYwMDAsNjcwMDAsODcwMDAsMTAzMDAwLDczMDAwLDExODAwMCw0NzAwMCwxNDAwMCw1OTAwMCwxMDEwMDAsMTIyMDAwLDc3MDAwLDEwNTAwMCwxMzkwMDAsODQwMDAsMTE2MDAwLDY0MDAwLDc0MDAwLDI1MDAwLDM2MDAwLDEwMDAsMjUwMDAsMTQ2MDAwLDE2MTAwMCwxMTkwMDAsNTYwMDAsNDAwMDAsNTAwMDAsOTAwMDAsNzcwMDAsNzkwMDAsMTI2MDAwLDE0MjAwMCw0NjAwMCwzMDAwLDE1MTAwMCw2NTAwMCwxMTcwMDAsOTUwMDAsMTMwMDAsNzkwMDAsODkwMDAsNTEwMDAsNjUwMDAsMTIwMDAwLDQ2MDAwLDgzMDAwLDUwMDAwLDc2MDAwLDczMDAwLDExODAwMCwxMzkwMDAsNzgwMDAsOTAwMCwyMTEwMDAsMjA3MDAwLDEyNDAwMCwxNDcwMDAsNjYwMDAsNDAwMCwxMjMwMDAsNTYwMDAsMTYwMDAsMTQ0MDAwLDk3MDAwLDEzNjAwMCwxNDIwMDAsMjUwMDAsMTUzMDAwLDE0NzAwMCwyMjIwMDAsMTEwMDAsODYwMDAsMTM3MDAwLDE0MDAwLDEwMDAsNTUwMDAsMTAwMCwxMzMwMDBdLCJvcHMiOjgxNjkwfV0sInVwZGF0ZWQiOiIyMDIzLTA0LTAzVDEzOjI3OjQyLjkzNloifQ%3D%3D).

Something like this code below:
```
export function isDelivered(packages: any) {
  if (packages == null || packages.length === 0) {
    return false
  }

  const isDelivered = packages.some(
    (pack: any) =>
      pack?.package?.type === OutputPackageType &&
      (!pack.package ||
        !pack.package.courierStatus ||
        pack.package.courierStatus.finished === false)
  )

  return isDelivered
}
```

What do you think?

Reply:
It looks great, Wagner!

If you add this code as a GitHub suggestion, I can apply it, and you will be the correct owner of it.

Then again, thanks!

Reply:
@andreaugusto I can't add my code as a suggestion because I can only add Github suggestions to changes, and my suggestion was to replace all the functions with a new code. Sorry 😢 
Could you get my suggestion and replace there?

Reply:
> I agree with using `some`, it makes a lot of sense, but I think this code is doing the opposite of what we want, right? I believe this `some` call would tell if the order is actually `notDelivered`. Also, is there a reason for this conditional expression to be different than `pack?.package?.courierStatus?.finished !== true`?

@batatavf2 
Yeah, my bad 😞 
You are right, I agree with you in both points about the logic using `some` and about `optional chaining`